### ISSUE TITLE
Explore sentence vectors and cosine difference as a potential method for ranking searchers

### ISSUE THREAD

**Issue by aarppe** [UAlbertaALTLab/cree-intelligent-dictionary#643]
METHOD
1. Make use of our existing 300-dimensional sentence vectors for CW content
   (undertaken by Atticus and Daniel). These exist as plain text and Rda source,
   and have been incorporated into python (Eddie). The one challenge here is
   that we'd need to add or revise vectors, when content is added to CW (or MD),
   but the algorigthm for that (averaging over individual sentence vectors in the
   English definitions) is trivial.

2. Make use of the original word2vec embeddings based on the GoogleNews corpus ([https://github.com/mmihaltz/word2vec-GoogleNews-vectors](https://github.com/mmihaltz/word2vec-GoogleNews-vectors))

```
>>> wv = KeyedVectors.load_word2vec_format('GoogleNews-vectors-negative300.bin.gz', binary=True)
>>> dim(wv)
>>> wv['sees',]
array([[ 0.15722656,  0.1171875 , -0.02526855, -0.07177734, -0.02404785,
...
```

3. We can use these word embeddings to create aggregate search vectors by averaging
   the individual word vectors of each lexical key word. Then we can calculate the
   cosine difference of the aggregated search key vector with the pre-calculated
   aggregate vectors for the CW entries, ranking CW entries with respect to the 
   search key words by decreasing cosine values (cosine=1 ~ perfect match vs.
   cosine=0 ~ no match). Examples below are quite promising.

**PRELIMINARY RESULTS USING R**

```
> sees_vs_CWv <- sapply(1:nrow(CW_verbs), function(i) lsa::cosine(sees, CW_verbs[i,]))
> rownames(CW_verbs)[order(sees_vs_CWv, decreasing=T)[1:10]]
 [1] "wâpiw"       "nahâpiw"     "wâpahtam"    "âpasâpahtam" "macâpisin"  
 [6] "nitawâpamêw" "miskwâpamêw" "wâpamêw"     "kîhkânâkêw"  "sâpwâpahtam"
```
```
> sees_witnesses_vs_CWv <- sapply(1:nrow(CW_verbs), function(i) lsa::cosine(apply(cbind(witnesses,sees),1,mean), CW_verbs[i,]))
> rownames(CW_verbs)[order(sees_witnesses_vs_CWv, decreasing=T)[1:10]]
 [1] "wâpiw"       "kîhkânâkêw"  "nahâpiw"     "miskwâpamêw" "wâpamêw"    
 [6] "wâpahtam"    "ayahcinâkêw" "pêtâpamêw"   "pêtisâpamêw" "nitawâpamêw"
```
```
> sees_clearly_vs_CWv <- sapply(1:nrow(CW_verbs), function(i) lsa::cosine(apply(cbind(sees, clearly),1,mean), CW_verbs[i,]))
> rownames(CW_verbs)[order(sees_clearly_vs_CWv, decreasing=T)[1:10]]
 [1] "nahâpiw"     "kîhkânam"    "nahâpahtam"  "pakaskinam"  "kîhkânâkêw" 
 [6] "kîhkânawêw"  "nahâpamêw"   "wâpiw"       "kîhkâyâstêw" "isinam"     
```
_nahâpiw_ ~ s/he sees clearly, s/he sees well; s/he has good vision [CW]
_kihkânam_ ~ s/he sees s.t. clearly, s/he recognizes it by sight [CW]
_nahâpahtam_ ~ s/he sees s.t. clearly, readily; s/he sees s.t. favourably

```
> works_together_vs_CWv <- sapply(1:nrow(CW_verbs), function(i) lsa::cosine(apply(cbind(works, together),1,mean), CW_verbs[i,]))
> rownames(CW_verbs)[order(works_together_vs_CWv, decreasing=T)[1:10]]
 [1] "wîtatoskêmêw"     "atoskêmakan"      "pimipayin"        "mâmawatoskêwak"  
 [5] "mâmawohkamwak"    "mâmawohkamâtowak" "nîswastâw"        "mâmawipayiwa"    
 [9] "mâmawêyatiwak"    "nîsowêwak"       
```

_wîtatoskêmêw_ ~ s/he works together with s.o., s/he has s.o. as his/her fellow worker
_atoskêmakan_ ~ it works, it functions
_pimipayin_ ~ it works, it goes, it functions; it runs, it runs along
_mâmawatoskêwak_ ~ they work together
_mâmawohkamwak_ ~ they work together on s.t. as a group ; they engage in a joint effort

**Comment by aarppe** [comment]
@andrewdotn Electronic food bits for thought about ranking search results.

**Comment by eddieantonio** [comment]
Implemented as of #786! 🎉
